src/main/java/desi/juan/email/api/Email.java

```diff
@@ -23,104 +25,245 @@
+   * @return the {@link LocalDateTime} this Email was received.
    */
-  Optional<LocalDateTime> getSentDate();
+  public abstract Optional<LocalDateTime> getReceivedDate();
 
   /**
-   * @return all the headers of this email message.
+   * @return an {@link EmailFlags} containing the flags set in the Email.
    */
-  Multimap<String, String> getHeaders();
+  public abstract EmailFlags getFlags();
 
   /**
-   * @return an {@link EmailFlags} object containing the flags setted in the email.
+   * The body of this Email
    */
-  EmailFlags getFlags();
+  private final EmailBody body;
 
   /**
-   * Get the body of the email, a text content of type "text/*";
+   * Get the body of this Email.
    *
-   * @return the body content of the email.
+   * @return the {@link EmailBody} of this Email.
+   */
+  public final EmailBody getBody() {
+    return body;
+  }
+
+  /**
+   * A {@link ImmutableList} with all the {@link EmailAttachment}s of this Email.
    */
-  EmailBody getBody();
+  private final ImmutableList<EmailAttachment> attachments;
 
   /**
-   * Get the attachments of the email, an empty list is returned when there are no attachments.
+   * Get the attachments of this Email. An empty {@link ImmutableList} is returned when there are no {@link EmailAttachment}s.
    *
-   * @return a {@link List} with the attachments of the email.
+   * @return a {@link ImmutableList} with the {@link EmailAttachment}s of this Email.
    */
-  List<EmailAttachment> getAttachments();
+  public final ImmutableList<EmailAttachment> getAttachments() {
+    return attachments;
+  }
+
+  /**
+   * The {@link Header}s of this Email.
+   */
+  private final ImmutableList<Header> headers;
+
+  /**
+   * @return all the @{link Header}s of this Email.
+   */
+  public final ImmutableList<Header> getHeaders() {
+    return headers;
+  }
+
+  /**
+   * Creates a new instance.
+   *
+   * @param subject
+   * @param sentDate
+   * @param from
+   * @param replyTo
+   * @param to
+   * @param cc
+   * @param bcc
+   * @param body
+   * @param attachments
+   * @param headers
+   */
+  protected Email(final String subject,
+                  final LocalDateTime sentDate,
+                  final ImmutableList<String> from,
+                  final ImmutableList<String> replyTo,
+                  final ImmutableList<String> to,
+                  final ImmutableList<String> cc,
+                  final ImmutableList<String> bcc,
+                  final EmailBody body,
+                  final ImmutableList<EmailAttachment> attachments,
+                  final ImmutableList<Header> headers) {
+
+    if (from.isEmpty()) {
+      throw new IllegalStateException(format("%s%s", CANNOT_BUILD, ERROR_FROM));
+    }
+
+    if (body == null) {
+      throw new IllegalStateException(format("%s%s", CANNOT_BUILD, ERROR_BODY));
+    }
+
+    // For an OutgoingEmail - TO, CC, or BCC must be present
+    if (getClass().equals(OutgoingEmail.class)) {
```

Comment:
this can be done in the constructor of OutgoingEmail, there is no need to check this.